src/main/java/com/minecolonies/coremod/colony/buildings/inventory/ItemStorageIndex.java

```diff
@@ -0,0 +1,117 @@
+package com.minecolonies.coremod.colony.buildings.inventory;
+
+import com.minecolonies.api.crafting.ItemStorage;
+import com.minecolonies.api.util.ItemStackUtils;
+import com.minecolonies.api.util.Tuple;
+import net.minecraft.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.*;
+import java.util.function.Predicate;
+
+/**
+ * Index which uses Items(ItemStorage) as key.
+ */
+public class ItemStorageIndex<Value> extends HashSetIndex<ItemStorage, Value>
```

Comment:
If this is your index for ItemStacks the index works backwards and is useless in its current implementation.
What you want is a mapping from ItemStorage (preferably directly ItemStack, since ItemStorage is not strict enough) to a Map of <somekind of lookup logic for IItemHandlers> and a list of slot indexes.

Cause your current logic is the same as a Map<ItemStorage, Integer>, which i can build easily enough from an Array or a List of ItemStorages. And as such provides no speed benefits.

Reply:
atm it is hierarchical and the index is split into two parts, one in the Building which maps ItemStorage to a set of racks, and an index in the racks which maps the itemstorage to the slots it is contained in

Reply:
I would merge this all together.